test/architecture/index.js

```diff
@@ -0,0 +1,212 @@
+const Product = require('../../src/collection/objects/product');
+
+const Resources = require('../../src/collection/resources/resources');
+const Resource = require('../../src/collection/objects/resource');
+
+const Subscriptions = require('../../src/collection/resources/subscriptions');
+const Subscription = require('../../src/collection/objects/subscription');
+
+const Users = require('../../src/collection/resources/users');
+const User = require('../../src/collection/objects/user');
+
+const architectureFile = require('./architecture.json');
+
+function getClassMethods(obj) {
+  const array = [];
+  const ignoreList = [
+    'constructor',
+    '__defineGetter__',
+    '__defineSetter__',
+    'hasOwnProperty',
+    '__lookupGetter__',
+    '__lookupSetter__',
+    'isPrototypeOf',
+    'propertyIsEnumerable',
+    'toString',
+    'valueOf',
+    'toLocaleString',
+
+    'paginate',
+    'query',
+    'ResConstructor',
+  ];
+  let names = Object.getOwnPropertyNames(obj);
+  names.forEach((name) => {
+    if (!ignoreList.includes(name) && typeof obj[name] === 'function') {
+      array.push(name);
+    }
+  });
+  let proto = Object.getPrototypeOf(obj);
+  while (proto) {
+    names = Object.getOwnPropertyNames(proto);
+    names.forEach((name) => {
+      if (!ignoreList.includes(name) && typeof obj[name] === 'function') {
+        array.push(name);
+      }
+    });
+    proto = Object.getPrototypeOf(proto);
+  }
+  return array;
+}
+
+
+const checkMatchingModels = (res, object) => {
+  getClassMethods(object).forEach((prop) => {
+    if (typeof object[prop] === 'function' && prop !== 'refresh') {
+      if (typeof object[prop] !== typeof res[prop]) {
+        console.error(prop, typeof object[prop], typeof res[prop]);
+      }
+      assert(typeof object[prop] === typeof res[prop]);
+    }
+  });
+};
+
+const areArraysEqual = (arr1, arr2) => {
+  for (let i = 0; i < arr1.length; i += 1) {
+    if (arr2.indexOf(arr1[i]) === -1) {
+      return false;
+    }
+  }
+  return true;
+};
+
+const checkArchitectureMatchesJsonFile = (id, res, object) => {
+  const model = architectureFile[id];
+
+  const objectFunctions = model.object.roles.user;
+  const resourceFunctions = model.resource.roles.user;
+
+  // We compare the actual function list with the expected function list.
+  assert(areArraysEqual(getClassMethods(object).sort(), objectFunctions.sort()));
+  assert(areArraysEqual(
+    getClassMethods(res).sort(), resourceFunctions.concat(objectFunctions).sort(),
+  ));
+
+  // We make sure no function is undefined.
+  objectFunctions.forEach((fn) => {
+    assert(object[fn] !== undefined);
+    // Object functions should be accessible from the resource equivalent object.
+    assert(fn === 'refresh' || res[fn] !== undefined);
+  });
+  resourceFunctions.forEach((fn) => {
+    assert(res[fn] !== undefined);
+  });
+};
+
+describe.only('Architectural tests', () => {
+  it('Should check architecture file correspong for AiModel', (done) => {
```

Comment:
several tests are broken, need to be fixed. Mostly due to changes in the available methods for resources